src/sentry_tracing.c

```diff
@@ -1,11 +1,245 @@
+            sentry_value_get_by_key_owned(parent, "span_id"));
+    }
+
+    return span;
+}
 
 sentry_value_t
-sentry__span_get_trace_context(sentry_value_t span)
+sentry__value_transaction_context_new(const char *name, const char *operation)
+{
+    sentry_value_t transaction_context
+        = sentry__value_new_span(sentry_value_new_null(), operation);
+
+    sentry_uuid_t trace_id = sentry_uuid_new_v4();
+    sentry_value_set_by_key(transaction_context, "trace_id",
+        sentry__value_new_internal_uuid(&trace_id));
+
+    sentry_value_set_by_key(
+        transaction_context, "transaction", sentry_value_new_string(name));
+
+    return transaction_context;
+}
+
+sentry_transaction_context_t *
+sentry_transaction_context_new(const char *name, const char *operation)
+{
+    sentry_transaction_context_t *tx_cxt
+        = SENTRY_MAKE(sentry_transaction_context_t);
+    if (!tx_cxt) {
+        return NULL;
+    }
+    memset(tx_cxt, 0, sizeof(sentry_transaction_context_t));
+
+    sentry_value_t inner
+        = sentry__value_transaction_context_new(name, operation);
+
+    if (sentry_value_is_null(inner)) {
+        return NULL;
+    }
+
+    tx_cxt->inner = inner;
+
+    return tx_cxt;
+}
+
+void
+sentry__transaction_context_free(sentry_transaction_context_t *tx_cxt)
+{
+    if (!tx_cxt) {
+        return;
+    }
+    if (sentry_value_refcount(tx_cxt->inner) <= 1) {
+        sentry_value_decref(tx_cxt->inner);
+        sentry_free(tx_cxt);
+    } else {
+        sentry_value_decref(tx_cxt->inner);
+    };
+}
+
+void
+sentry_transaction_context_set_name(
+    sentry_transaction_context_t *tx_cxt, const char *name)
 {
-    if (sentry_value_is_null(span)
-        || sentry_value_is_null(sentry_value_get_by_key(span, "trace_id"))
-        || sentry_value_is_null(sentry_value_get_by_key(span, "span_id"))) {
+    sentry_value_set_by_key(
+        tx_cxt->inner, "transaction", sentry_value_new_string(name));
+}
+
+void
+sentry_transaction_context_set_operation(
+    sentry_transaction_context_t *tx_cxt, const char *operation)
+{
+    sentry_value_set_by_key(
+        tx_cxt->inner, "op", sentry_value_new_string(operation));
+}
+
+void
+sentry_transaction_context_set_sampled(
+    sentry_transaction_context_t *tx_cxt, int sampled)
+{
+    sentry_value_set_by_key(
+        tx_cxt->inner, "sampled", sentry_value_new_bool(sampled));
+}
+
+void
+sentry_transaction_context_remove_sampled(sentry_transaction_context_t *tx_cxt)
+{
+    sentry_value_remove_by_key(tx_cxt->inner, "sampled");
+}
+
+sentry_transaction_t *
+sentry__transaction_new(sentry_value_t inner)
+{
+    if (sentry_value_is_null(inner)) {
+        return NULL;
+    }
+
+    sentry_transaction_t *tx = SENTRY_MAKE(sentry_transaction_t);
+    if (!tx) {
+        return NULL;
+    }
+    memset(tx, 0, sizeof(sentry_transaction_t));
+
+    tx->inner = inner;
+
+    return tx;
+}
+
+void
+sentry__transaction_incref(sentry_transaction_t *tx)
+{
+    sentry_value_incref(tx->inner);
+}
```

Comment:
will probably need some clarification on this; is the idea to remove the ability to create multiple references to a transaction? or is it that the implementation of "reference counting" on the transaction adds an unnecessary layer of abstraction over `sentry_value_t`'s reference counting?

if the intent is to prevent any reference counting at all, we would have to figure out how to let the scope obtain up to date information on a transaction without resorting to refcounting. one idea could be to have the scope contain some separately allocated copy of the transaction's data.

Reply:
my rough thinking was to avoid needing to reference count the transaction entirely.  Instead store a pointer in the transaction to the scope, which can be NULL if the transaction is not bound to the scope.  Then whenever the transaction updates something which needs to be represented on the scope check the pointer and go update the scope.

of course it might be that my rough thinking is just wrong :)